File: MPChartLib/src/main/java/com/github/mikephil/charting/utils/FileUtils.java

package com.github.mikephil.charting.utils;

import android.content.res.AssetManager;
import android.util.Log;

import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.Entry;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Utilities class for interacting with the assets and the devices storage to
 * load and save DataSet objects from and to .txt files.
 *
 * @author Philipp Jahoda
 */
public class FileUtils {

	private static final String LOG = "Chart-FileUtils";

	/**
	 * Loads an array of Entries from a textfile from the assets folder.
	 *
	 * @param path the name of the file in the assets folder (+ path if needed)
	 */
	public static List<Entry> loadEntriesFromAssets(AssetManager am, String path) {

		List<Entry> entries = new ArrayList<>();

		try (BufferedReader reader = new BufferedReader(
				new InputStreamReader(am.open(path), StandardCharsets.UTF_8))) {

			String line = reader.readLine();

			while (line != null) {
				// process line
				String[] split = line.split("#");

				if (split.length <= 2) {
					entries.add(new Entry(Float.parseFloat(split[1]), Float.parseFloat(split[0])));
				} else {

					float[] vals = new float[split.length - 1];

					for (int i = 0; i < vals.length; i++) {
						vals[i] = Float.parseFloat(split[i]);
					}

					entries.add(new BarEntry(Integer.parseInt(split[split.length - 1]), vals));
				}
				line = reader.readLine();
			}
		} catch (IOException e) {
			Log.e(LOG, e.toString());

		}

		return entries;
	}

	public static List<BarEntry> loadBarEntriesFromAssets(AssetManager am, String path) {

		List<BarEntry> entries = new ArrayList<>();

		try (BufferedReader reader = new BufferedReader(
				new InputStreamReader(am.open(path), StandardCharsets.UTF_8))) {

			String line = reader.readLine();

			while (line != null) {
				// process line
				String[] split = line.split("#");

				entries.add(new BarEntry(Float.parseFloat(split[1]), Float.parseFloat(split[0])));

				line = reader.readLine();
			}
		} catch (IOException e) {
			Log.e(LOG, e.toString());

		}

		return entries;
	}
}

File: MPChartLib/src/main/java/com/github/mikephil/charting/utils/FileUtils.kt
package com.github.mikephil.charting.utils

import android.content.res.AssetManager
import android.util.Log
import com.github.mikephil.charting.data.BarEntry
import com.github.mikephil.charting.data.Entry
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

/**
 * Utilities class for interacting with the assets and the devices storage to load and save DataSet objects from and to .txt files.
 */
object FileUtils {
    private const val LOG = "Chart-FileUtils"

    /**
     * Loads an array of Entries from a textfile from the assets folder.
     *
     * @param path the name of the file in the assets folder (+ path if needed)
     */
    fun loadEntriesFromAssets(am: AssetManager, path: String): MutableList<Entry> {
        val entries: MutableList<Entry> = ArrayList()

        try {
            BufferedReader(
                InputStreamReader(am.open(path), StandardCharsets.UTF_8)
            ).use { reader ->
                var line = reader.readLine()
                while (line != null) {
                    // process line
                    val split: Array<String?> = line.split("#".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

                    if (split.size <= 2) {
                        entries.add(Entry(split[1]!!.toFloat(), split[0]!!.toFloat()))
                    } else {
                        val vals = FloatArray(split.size - 1)

                        for (i in vals.indices) {
                            vals[i] = split[i]!!.toFloat()
                        }

                        entries.add(BarEntry(split[split.size - 1]!!.toInt().toFloat(), vals))
                    }
                    line = reader.readLine()
                }
            }
        } catch (e: IOException) {
            Log.e(LOG, e.toString())
        }

        return entries
    }

    fun loadBarEntriesFromAssets(am: AssetManager, path: String): MutableList<BarEntry> {
        val entries: MutableList<BarEntry> = ArrayList()

        try {
            BufferedReader(InputStreamReader(am.open(path), StandardCharsets.UTF_8)).use { reader ->
                var line = reader.readLine()
                while (line != null) {
                    val split: Array<String?> = line.split("#".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    entries.add(BarEntry(split[1]!!.toFloat(), split[0]!!.toFloat()))
                    line = reader.readLine()
                }
            }
        } catch (e: IOException) {
            Log.e(LOG, e.toString())
        }

        return entries
    }
}
